src/yetibot/core/models/default_command.clj

```diff
@@ -1,35 +1,52 @@
 (ns yetibot.core.models.default-command
   "Determine which command to fallback to and whether fallback is enabled"
-  (:require
-    [clojure.string :refer [blank?]]
-    [clojure.spec.alpha :as s]
-    [yetibot.core.config :refer [get-config]]))
+  (:require [clojure.spec.alpha :as s]
+            [yetibot.core.config :refer [get-config]]
+            [clojure.string :refer [blank?]]))
 
 (s/def ::config any?)
 
-(defn configured-default-command []
-  (or
-    (:value (get-config ::config [:default :command]))
-    "help"))
+(defn configured-default-command
+  "Gets the default command, as defined by the instance config.
+   Arity/2 allows for passing in custom config to compare against,
+   mainly used for testing."
+  ([] (configured-default-command (get-config ::config
+                                              [:default :command])))
+  ([cfg] (get cfg :value "help")))
+
+(comment
+  (configured-default-command {:value "findme"})
+  (configured-default-command {})
+  )
 
 (s/def ::text string?)
 
 (defn fallback-help-text-override
-  "Optional config, may be nil"
-  []
-  (:value (get-config ::text [:command :fallback :help :text])))
+  "Optional config for fallback help text. May be nil. Arity/2 allows for passing
+   in custom config to compare against, mainly used for testing."
+  ([] (fallback-help-text-override (get-config ::text
+                                               [:command :fallback :help :text])))
+  ([cfg] (:value cfg)))
 
 (s/def ::fallback-commands-enabled-config string?)
 
 (defn fallback-enabled?
   "Determine whether fallback commands are enabled when user enters a command
    that doesn't exist. Default is true.
+   In the future this may be channel specific but for now it is global.
+   Arity/2 allows for passing in custom config to compare against, mainly used
+   for testing."
+  ([] (fallback-enabled? (get-config ::fallback-commands-enabled-config
```

Comment:
hrmmm -- all good points .. my approach has been simply to make it as "pure" as possible -- which in this case means passing in a config map vs deriving it from `(get-config)` .. that feels "clean" to me and is easy ..

that said, i do feel like we are missing out on NOT testing what is returned from `(get-config)` -- but to be honest, my limited brain was not able to figure out how to mock said config -- and so i went with the "pure" (increased arity) approach ..

is there any way you can show me an example of how i might mock that `(get-config)` ?? i see the link to "background prereqs" -- but it is not clicking for me ..

Reply:
Agree the docs are confusing. Yes I'll try to get an example out today!

Reply:
👍 no rush .. i am more than happy to take this approach, as i think it will reduce any potential confusion around why the continued increase of arity .. that said, i really do like the purity of what i was testing ..

i did try to implement how i think it should work, but keep getting errors

```clojure
 (fact
  (against-background (get-config ::config [:default :command]) => {:value "somecommand"})
  (dc/configured-default-command) => "somecommand")
 (fact
  (against-background (get-config ::config [:default :command]) => {:error :not-found})
  (dc/configured-default-command) => "help")
```

```
FAIL at (default_command.clj:16)
You never said #'get-config would be called with these arguments:
    [:yetibot.core.models.default-command/config [:default :command]]

FAIL about configured-default-command at (default_command.clj:17)
Expected:
"somecommand"
Actual:
"help"
Diffs: strings have 1 difference (0% similarity)
                expected: "(somecommand)"
                actual:   "(help-------)"

FAIL at (default_command.clj:19)
You never said #'get-config would be called with these arguments:
    [:yetibot.core.models.default-command/config [:default :command]]
```

Reply:
this is interesting .. when i create a new function that contains the `(get-config)` function and use it instead of calling `(get-config)` directly in the `(configured-default-command)` function -- my mocks seem to work ..

for example:
```clojure
;; the source .. removed doc strings for brevity
(defn def-cmd [] (get-config ::config [:default :command]))

(defn configured-default-command
  ([] (get (def-cmd) :value "help")))
```

```clojure
;; the tests
 (fact
  (against-background (dc/def-cmd) => {:value "somecommand"})
  (dc/configured-default-command) => "somecommand")
 (fact
  (against-background (dc/def-cmd) => {:error :not-found})
  (dc/configured-default-command) => "help")
```

all tests pass as defined .. is this the correct path ?? or is there some other preferred mechanism ??

thanks,
greg

Reply:
I was thinking more like this, using `provided` for the config itself:

```clojure
(ns yetibot.core.test.models.default-command
  (:require [yetibot.core.models.default-command :as dc]
            [yetibot.core.config :refer [get-config]]
            [midje.sweet :refer [=> provided anything fact facts throws]]))

(facts
 "about configured-default-command"
 (fact
  "returns a custom default command"
  (dc/configured-default-command) => "findme"
  (provided (get-config anything [:default :command]) => {:value "findme"}))
 ;; ...
 )
```

See [test/commands/history](https://github.com/yetibot/core/blob/master/test/yetibot/core/test/commands/history.clj#L30) for a bunch of other examples of `provided`.

I love pure fns, but I'd argue that making a separate arity for the purpose of testing while the original used by app is still impure isn't much of an improvement in terms of making the code pure. I guess thinking about it a little more I probably wouldn't want this 2-arity pattern sprinkled throughout the entire code base. If we want increased purity in general I'm always up for thinking about that.

Reply:
👍 thanks for the example and guidance .. i am going to close this PR and open a new one once i feel like i have solid examples of mocking in place ..
your comments about the original function signature remaining impure is very true, and so the net effect is little improvement in purity, but increased LoC side-effects << which isn't awesome ..
maybe in the future, we can ponder on how we might want to increase purity in key areas, so it is easier for folks like myself to more easily grasp the codebase ..
OK then -- expect a new PR in the next day or so ..

Reply:
Amazing, thanks @gkspranger !!